### PR TITLE
[AMDGPU][AsmParser] Support structured HWREG operands.

### PR DESCRIPTION
**Pull request by kosarev**
Symbolic values are to be supported separately.

### CODE REVIEW

**Review by arsenm — APPROVED** [review]
I assume this is a syntax sp3 supports?

**Comment by jayfoad** [comment]
> I assume this is a syntax sp3 supports?

Yes. I believe SP3 started supporting it in the GFX11 timeframe, but it has backported the syntax to instructions that predate GFX11 like getreg/setreg, and accepts the syntax even when assembling for a pre-GFX11 ASIC.

File: llvm/test/MC/AMDGPU/sopk-err.s [hunk 1]
```diff
@@ -5,48 +5,122 @@
...
+// RUN: not llvm-mc -triple=amdgcn -mcpu=gfx1010 %s 2>&1 | FileCheck -check-prefixes=GCN,GFX10-ERR --implicit-check-not=error: --strict-whitespace %s
+// RUN: not llvm-mc -triple=amdgcn -mcpu=gfx1100 %s 2>&1 | FileCheck -check-prefixes=GCN,GFX11-ERR --implicit-check-not=error: --strict-whitespace %s
 
 s_setreg_b32  0x1f803, s2
 // GCN: :[[@LINE-1]]:{{[0-9]+}}: error: invalid immediate: only 16-bit values are legal
+// GCN-NEXT: {{^}}s_setreg_b32  0x1f803, s2
+// GCN-NEXT: {{^}}              ^
 
 s_setreg_b32  typo(0x40), s2
 // GCN: :[[@LINE-1]]:{{[0-9]+}}: error: expected a hwreg macro or an absolute expression
```
Comment (jayfoad) [inline llvm/test/MC/AMDGPU/sopk-err.s:21]: Nit: "... or a structured initializer"?

File: llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp [hunk 2]
```diff
@@ -7730,6 +7741,44 @@ AMDGPUAsmParser::getConstLoc(const OperandVector &Operands) const {
...
 }
 
+ParseStatus
+AMDGPUAsmParser::parseStructuredOpFields(ArrayRef<StructuredOpField *> Fields) {
+  if (!trySkipToken(AsmToken::LCurly))
+    return ParseStatus::NoMatch;
+
+  bool First = true;
+  while (!trySkipToken(AsmToken::RCurly)) {
+    if (!First && !skipToken(AsmToken::Comma, "comma expected"))
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:7751]: "expect comma or right brace"?

File: llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp [hunk 3]
```diff
@@ -7730,6 +7741,44 @@ AMDGPUAsmParser::getConstLoc(const OperandVector &Operands) const {
...
+    if (!skipToken(AsmToken::Identifier, "field name expected") ||
+        !skipToken(AsmToken::Colon, "colon expected"))
+      return ParseStatus::Failure;
+
+    auto I =
+        find_if(Fields, [Id](StructuredOpField *F) { return F->Id == Id; });
+    if (I == Fields.end())
+      return Error(IdLoc, "unknown field");
+
+    // TODO: Support symbolic values.
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:7765]: Diagnose duplicate fields here, like `{id:1,id:2}`? What does SP3 do?

**Review by jayfoad — COMMENTED** [review]
LGTM with some nits

File: llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp (hunk 2 above)
Comment (kosarev) [reply llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp]: Done.

File: llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp (hunk 3 above)
Comment (kosarev) [reply llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:7770]: SP3 allows duplicates and takes the value specified last.

File: llvm/test/MC/AMDGPU/sopk-err.s (hunk 1 above)
Comment (kosarev) [reply llvm/test/MC/AMDGPU/sopk-err.s]: Done. I called it 'structured immediate' as 'initializer' looks a bit alien to the context.

**Comment by kosarev** [comment]
Updated as suggested and rebased.

**Review by jayfoad — APPROVED** [review]
LGTM, thanks!

**Pull request merged**